lib/bloc/workout/workout.dart

```diff
@@ -0,0 +1,90 @@
+import 'package:event_bloc/event_bloc.dart';
+import 'package:search_me_up/search_me_up.dart';
+import 'package:simple_exercise_builder/model/workout.dart';
+import 'package:simple_exercise_builder/repository/workout_repo.dart';
+
+class WorkoutBloc extends Bloc {
+  @override
+  final BlocEventChannel eventChannel;
+  final WorkoutRepository repo;
+
+  late final _searchMeUp = SearchMeUp<Workout>(
+    DefaultSearchMeUpDelegate(
+      converters: [
+        (workout) => [workout.name!]
+      ],
+    ),
+  );
+
+  WorkoutBloc({BlocEventChannel? parentChannel, required this.repo})
+      : eventChannel = BlocEventChannel(parentChannel);
+
+  List<int> workoutList = [];
+  Map<int, Workout> workoutMap = {};
+  String? searchTerm;
+
+  /// This will load the workouts from the [repo]
+  void loadWorkouts() async {
+    final workouts = await repo.retrieveWorkouts();
+    workoutMap.clear();
+    workouts.forEach((element) => workoutMap[element.id!] = element);
+    _generateWorkoutList();
+    updateBloc();
+  }
+
+  void search(String search) {
+    final actualSearch = search.trim().isEmpty ? null : search.trim();
+    if (actualSearch == searchTerm) {
+      return;
+    }
+
+    searchTerm = actualSearch;
+    _generateWorkoutList();
+    updateBloc();
+  }
+
+  void addNewWorkout() async {
+    await repo.addWorkout(
+      Workout(
+        name: 'New Workout',
+        userMade: true,
+      ),
+    );
+
+    loadWorkouts();
+  }
+
+  void deleteWorkout(Workout workout) async {
+    await repo.deleteWorkout(workout);
+    loadWorkouts();
+  }
+
+  void duplicateWorkout(Workout workout) async {
+    final clone = Workout();
+    clone.loadFromMap(workout.toMap());
+    clone.name = 'Copy of ${clone.name}';
```

Comment:
I plan that the repository/database will take care of the setting of ID when adding a new workout. I could manually set the id to null to make it more apparent.

Also, I think toMap  should not save the ID if we will be using the map for deep links/sharing. This is because the ID might already be taken on the side of the receiver.

Reply:
Set the id to null at 104f62b

Reply:
Deep linking won't use toMap. It'll use a more specific command. Still needs to be designed though. Hahahaha.

ToMap and loadFromMap are more for putting it into Hive and extracting it from it.